### PR TITLE
[Feed] Implement `Storage` logic (2/2)

### LINKED ISSUE #1375: [Feed] Populate logic in `RUFAS/routines/feed_storage/storage.py`
Populate the logics of the following methods in `RUFAS/routines/feed_storage/storage.py`:

- [ ] `process_degradations`
- [ ] `calculate_dry_matter_loss_to_gas`
- [ ] `calculate_dry_matter_loss_to_effluent`
- [ ] `calculate_protein_degradation`
- [ ] `calculate_protein_degradation`
- [ ] `calculate_bale_density`
- [ ] `recalculate_nutrient_fractions`

[Design doc is here](https://docs.google.com/document/d/1C86lch2dkYBERR-qLDmE7L7n1gYQjETrIKUyrQpWrPs/edit)

* For questions, reach out to SME: @KevinP-B and/or SWE @PooyaHekmati . Their reviews are needed for any linked PRs.
* Size the PRs reasonably, not all the methods have to be implemented in a single PR.
* docstrings, function signatures, and unit tests need to be updated accordingly.
* I suggest reading and understanding the design doc before attempting to write the code.

### PR DESCRIPTION
**Pull request by Isgotkowitz**
Adds additional logic to the `Storage` class to update mass attributes after loss and record the current aggregated state of stored crops.

### Context

Issue(s) closed by this pull request: closes #1375 

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.

How big are the changes in the PR? Would you nominate it for a major version upgrade?
- [ ] Yes
- [x] No

### What

The logic for 3 methods have been added:
- `reset_mass_attributes_after_loss`
- `record_stored_crops`
- `_get_total_nutritive_amount`

These methods add the rest of the logic that needs to be present in `Storage`.

Note that this PR also modifies `record_stored_crops` to not record the gaseous dry matter loss - this makes the code more reusable because not all storage types lose mass to dry matter loss.

### Why

When dry matter mass is lost it is critical to make sure that loss is accounted for, which is what the `reset_mass_attributes_after_loss` does. It is also a requirement that outputs from the Feed Storage module are logged daily, and `record_stored_crops` helps achieve this by taking a daily snapshot of the total amounts of mass and nutrients in a storage.

### How

To recalculate dry matter, `reset_mass_attributes` takes in a stored crop and the amount of dry matter it lost, recalculates the dry and fresh mass, then sets the fresh mass and recalculates and sets the percentage of the fresh mass that is dry matter.
`record_stored_crops` goes through a number of different attributes in harvested crop, combines the same attributes from a crop into a total, and reports it. It uses `_get_total_nutritive_amount` as a helper to find the total amounts of specific nutrients held in storage.

### Test plan

All methods have had unit tests written. The coverage of `Storage` is at 99%, with the only uncovered line being a `pass` that is outside the scope of this PR.

### Open Questions
- Is `record_stored_crops` the best way to capture the state of `Storage`? It provides a very coarse grained look at what is going on in the stored crops, which is both a benefit and a drawback.
  - An alternative or supplement would be to record information on a stored crop by stored crop basis, but one of the main challenges of this is how to distinguish between stored crops?

### CODE REVIEW

**Review by KevinP-B — COMMENTED** [review]
Looks good to me.

Regarding open issue:

I think this would have to be addressed by long-form conversation to weigh the relative pros and cons.

**Review by KevinP-B — APPROVED** [review]
Looks sound. Test outputs look good.

File: RUFAS/routines/feed_storage/storage.py [hunk 1]
```diff
@@ -200,19 +207,72 @@ def reset_mass_attributes_after_loss(self, crop: HarvestedCrop, dry_matter_loss:
             Amount of dry matter the crop lost on the current day in kg.
 
         """
-        pass
-
-    def record_stored_crops(self, gaseous_dry_matter_loss: float) -> None:
+        new_dry_matter_mass = crop.dry_matter_mass - dry_matter_loss
+        crop.fresh_mass -= dry_matter_loss
+        if crop.fresh_mass == 0.0:
```
Comment (ltp23) [inline RUFAS/routines/feed_storage/storage.py:217]: I think doing a direct comparison may not work all the time. Maybe calculating the difference from 0 and check for tolerance level.

File: RUFAS/routines/feed_storage/storage.py [hunk 2]
```diff
@@ -200,19 +207,72 @@ def reset_mass_attributes_after_loss(self, crop: HarvestedCrop, dry_matter_loss:
...
+            crop.dry_matter_percentage = 0.0
+            return
+        crop.dry_matter_percentage = new_dry_matter_mass / crop.fresh_mass * GeneralConstants.FRACTION_TO_PERCENTAGE
+
+    def record_stored_crops(self) -> None:
         """
         Records the total mass and nutrient amounts held in storage.
+        """
+        info_map = {"class": self.__class__.__name__, "function": self.record_stored_crops.__name__, "units": "kg"}
```
Comment (ltp23) [inline RUFAS/routines/feed_storage/storage.py:226]: The units could use MeasureUnits.KILOGRAMS.

File: tests/test_feed_storage_module/test_storage.py [hunk 3]
```diff
@@ -97,6 +97,11 @@ def test_process_degradations(
     """
     Test the process_degradations method of the Storage class.
     """
+    expected_info_map = {
+        "class": storage.__class__.__name__,
+        "function": storage.process_degradations.__name__,
+        "units": "kg",
```
Comment (ltp23) [inline tests/test_feed_storage_module/test_storage.py]: MeasurementUnits.KILOGRAMS

**Review by ltp23 — APPROVED** [review]
Overall, it looks good to me. Thanks Ed.

**Pull request merged**